File: Anul 2/Semestrul 2/MPP/labs/Tema 5-proiect Server Csharp-Client Java/proiectSCcsharp/protobuf/ProtoClientProxy.cs
using proiectSCcsharp.services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization.Formatters.Binary;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;
using Chat.Protocol;
using Google.Protobuf;

namespace protobuf
{
    internal class ProtoClientProxy:IService
    {
        private readonly string host;
        private readonly int port;
        private IObserver employeeObserver;
        private NetworkStream stream;
        private IFormatter formatter;
        private TcpClient connection;
        private readonly Queue<AppResponse> responses;
        private volatile bool finished;
        private EventWaitHandle waitHandle;

        public ProtoClientProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            responses = new Queue<AppResponse>();
        }

        private void CloseConnection()
        {
            finished = true;
            try
            {
                stream.Close();
                connection.Close();
                waitHandle.Close();
                employeeObserver = null;
                Console.WriteLine("Connection closed!");
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private void SendRequest(AppRequest request)
        {
            try
            {
                request.WriteDelimitedTo(stream);
                stream.Flush();
            }
            catch (Exception e)
            {
                throw new MyException("Error sending object " + e);
            }
        }

        private AppResponse ReadResponse()
        {
            AppResponse response = null;
            try
            {
                waitHandle.WaitOne();
                lock (responses)
                {
                    response = responses.Dequeue();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
            return response;
        }

        private void InitializeConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                finished = false;
                waitHandle = new AutoResetEvent(false);
                StartReader();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private void StartReader()
        {
            var tw = new Thread(Run);
            tw.Start();
        }

        private void HandleUpdate(AppResponse response)
        {
            if (response.Type != AppResponse.Types.Type.SoldBilet) return;
            var ticket = ProtoUtils.GetBilet(response);
            Console.WriteLine("Ticket sold " + ticket);
            try
            {
                employeeObserver.BiletAdded(ticket);
            }
            catch (MyException e)
            {
                Console.WriteLine("Error handle update: " + e);
            }
        }

        private static bool IsUpdate(AppResponse response)
        {
            return response.Type == AppResponse.Types.Type.SoldBilet;
        }

        public proiectSCcsharp.domain.Angajat login(proiectSCcsharp.domain.Angajat employee, IObserver employeeObserver)
        {
            InitializeConnection();
            Console.WriteLine("Login request ...");
            var req = ProtoUtils.createLoginRequest(employee);
            SendRequest(req);
            var response = ReadResponse();
            switch (response.Type)
            {
                case AppResponse.Types.Type.Ok:
                    this.employeeObserver = employeeObserver;
                    Console.WriteLine("Logged in");
                    return ProtoUtils.GetAngajat(response);
                case AppResponse.Types.Type.Error:
                    {
                        var err = ProtoUtils.getError(response);
                        Console.WriteLine("Error logging in" + err);
                        CloseConnection();
                        throw new MyException(err);
                    }
                default:
                    return null;
            }
        }

        public void logout(proiectSCcsharp.domain.Angajat employee)
        {
            var req = ProtoUtils.createLogoutRequest(employee);
            SendRequest(req);
            var response = ReadResponse();
            CloseConnection();
            if (response.Type == AppResponse.Types.Type.Error)
            {
                var err = ProtoUtils.getError(response);
                Console.WriteLine("Error logging out" + err);
                throw new MyException(err);
            }
            Console.WriteLine("Logged out");
        }

        public void addBilet(proiectSCcsharp.domain.Bilet bilet)
        {
            var req = ProtoUtils.createAddBiletRequest(bilet);
            SendRequest(req);
            var response = ReadResponse();
            if (response.Type == AppResponse.Types.Type.Error)
            {
                var err = ProtoUtils.getError(response);
                Console.WriteLine("Error adding ticket" + err);
                throw new MyException(err);
            }
            Console.WriteLine("Ticket added");
        }

        public List<proiectSCcsharp.domain.Spectacol> getAllSpectacole()
        {
            var req = ProtoUtils.createGetAllSpectacoleRequest();
            SendRequest(req);
            var response = ReadResponse();
            if (response.Type == AppResponse.Types.Type.Error)
            {
                var err = ProtoUtils.getError(response);
                Console.WriteLine("Error getting shows" + err);
                throw new MyException(err);
            }
            Console.WriteLine("Got shows");
            return ProtoUtils.getAllSpectacole(response);
        }

        public proiectSCcsharp.domain.Spectacol getSpectacol(long id)
        {
            var req = ProtoUtils.createGetSpectacolRequest(id);
            SendRequest(req);
            var response = ReadResponse();
            if (response.Type == AppResponse.Types.Type.Error)
            {
                var err = ProtoUtils.getError(response);
                Console.WriteLine("Error finding spectacol" + err);
                throw new MyException(err);
            }
            Console.WriteLine("Found spectacol");
            return ProtoUtils.GetSpectacol(response);
        }

        protected virtual void Run()
        {
            while (!finished)
            {
                try
                {
                    var response = AppResponse.Parser.ParseDelimitedFrom(stream);
                    Console.WriteLine("response received " + response);
                    if (IsUpdate((AppResponse)response))
                    {
                        HandleUpdate((AppResponse)response);
                    }
                    else
                    {
                        lock (responses)
                        {
                            responses.Enqueue((AppResponse)response);
                        }
                        waitHandle.Set();
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine("Reading error " + e);
                }
            }
        }
    }
}

File: Anul 2/Semestrul 2/MPP/labs/Tema 5-proiect Server Csharp-Client Java/proiectServerClientJava/Networking/src/main/java/proiectServerClientJava/network/protobuffprotocol/ProtoClientProxy.java
package proiectServerClientJava.network.protobuffprotocol;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import proiectServerClientJava.domain.*;
import proiectServerClientJava.service.*;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ProtoClientProxy implements IService {
    private String host;
      private int port;

      private IObserver client;

      private InputStream input;
      private OutputStream output;
      private Socket connection;

      private BlockingQueue<ClientProtobufs.AppResponse> qresponses;
      private volatile boolean finished;
    private static final Logger logger = LogManager.getLogger();

    public ProtoClientProxy(String host, int port) {
          this.host = host;
          this.port = port;
          qresponses=new LinkedBlockingQueue<ClientProtobufs.AppResponse>();
      }

      public Angajat login(Angajat employee, IObserver client) throws MyException {
          initializeConnection();
          System.out.println("Login request ...");
          sendRequest(ProtoUtils.createLoginRequest(employee));
          ClientProtobufs.AppResponse response=readResponse();
          if (response.getType()==ClientProtobufs.AppResponse.Type.Ok){
              this.client=client;
              return ProtoUtils.getAngajat(response);
          }
          if (response.getType()==ClientProtobufs.AppResponse.Type.Error){
              String errorText=ProtoUtils.getError(response);
              closeConnection();
              throw new MyException(errorText);
          }
          return null;
      }

    public void addBilet(Bilet bilet) throws MyException {
        sendRequest(ProtoUtils.createAddBiletRequest(bilet));
        ClientProtobufs.AppResponse response=readResponse();
        if (response.getType()==ClientProtobufs.AppResponse.Type.Error){
            String errorText=ProtoUtils.getError(response);
            throw new MyException(errorText);
        }
        else {
            logger.info("Ticket added");
            System.out.println("Ticket added");
        }
    }

    public void logout(Angajat employee, IObserver client) throws MyException {
        sendRequest(ProtoUtils.createLogoutRequest(employee));
        ClientProtobufs.AppResponse response=readResponse();
        closeConnection();
        if (response.getType()==ClientProtobufs.AppResponse.Type.Error){
            String errorText=ProtoUtils.getError(response);
            logger.error("Error logging out" + errorText);
            System.out.println("Error logging out" + errorText);
            throw new MyException(errorText);
        }
        else {
            logger.info("Logged out");
            System.out.println("Logged out");
        }
    }
    public Spectacol getSpectacol(Long id) throws MyException {
        sendRequest(ProtoUtils.createGetSpectacolRequest(id));
        ClientProtobufs.AppResponse response=readResponse();
        if (response.getType()==ClientProtobufs.AppResponse.Type.Error){
            String errorText=ProtoUtils.getError(response);
            throw new MyException(errorText);
        }
        Spectacol spectacol = ProtoUtils.getSpectacol(response);
        return spectacol;
    }

    public Iterable<Spectacol> getAllSpectacole() throws MyException {
        sendRequest(ProtoUtils.createGetAllSpectacoleRequest());
        ClientProtobufs.AppResponse response=readResponse();
        if (response.getType()==ClientProtobufs.AppResponse.Type.Error){
            String errorText=ProtoUtils.getError(response);
            throw new MyException(errorText);
        }
        Iterable<Spectacol> spectacole=ProtoUtils.getAllSpectacole(response);
        return spectacole;
    }

      private void closeConnection() {
          finished=true;
          try {
              input.close();
              output.close();
              connection.close();
              client=null;
              logger.info("Closed connection");
              System.out.println("Closed connection");
          } catch (IOException e) {
              logger.error("Error closing connection" + e);
              System.out.println("Error closing connection" + e);
          }

      }

      private void sendRequest(ClientProtobufs.AppRequest request)throws MyException{
          try {
              System.out.println("Sending request ..."+request);
              logger.info("Sending request ..." + request);
              request.writeDelimitedTo(output);
              output.flush();
              System.out.println("Request sent.");
              logger.info("Request sent.");
          } catch (IOException e) {
              logger.error("Error sending request " + e);
              throw new MyException("Error sending object "+e);
          }

      }

      private ClientProtobufs.AppResponse readResponse() throws MyException{
          ClientProtobufs.AppResponse response=null;
          try {
              response = qresponses.take();
              logger.info("Read response " + response);
          } catch (InterruptedException e) {
              logger.error("Error reading response " + e);
              throw new MyException("Error reading response " + e);
          }
          return response;
      }
      private void initializeConnection() throws MyException{
           try {
              connection=new Socket(host,port);
              output=connection.getOutputStream();
              input=connection.getInputStream();
              finished=false;
              startReader();
          } catch (IOException e) {
              e.printStackTrace();
          }
      }
      private void startReader(){
          Thread tw=new Thread(new ReaderThread());
          tw.start();
      }


      private void handleUpdate(ClientProtobufs.AppResponse updateResponse){
          if (updateResponse.getType() == ClientProtobufs.AppResponse.Type.SoldBilet) {
              Bilet ticket = ProtoUtils.getBilet(updateResponse);
              logger.info("Ticket sold " + ticket);
              System.out.println("Ticket sold " + ticket);
              try {
                  client.biletAdded(ticket);
              } catch (MyException e) {
                  logger.error("Error handle update: " + e);
                  System.out.println("Error handle update: " + e);
              }
          }

      }
      private class ReaderThread implements Runnable{
          public void run() {
              while (!finished) {
                  try {
                      ClientProtobufs.AppResponse response = ClientProtobufs.AppResponse.parseDelimitedFrom(input);
                      System.out.println("response received " + response);
                      if (response != null) {
                          if (isUpdateResponse(response.getType())) {
                              handleUpdate(response);
                          } else {
                              try {
                                  qresponses.put(response);
                              } catch (InterruptedException e) {
                                  logger.error(e);
                              }
                          }
                      }
                  } catch (IOException e) {
                      System.out.println("Reading error " + e);
                  }
              }
          }
      }

    private boolean isUpdateResponse(ClientProtobufs.AppResponse.Type type){
        return type == ClientProtobufs.AppResponse.Type.SoldBilet;

    }
}
